### DISCUSSION TITLE
getMarkerFeatures() all groups vs specific group

### DISCUSSION THREAD

**Discussion by wuv21** [GreenleafLab/ArchR#1448] [Questions / Documentation]
Hi! Just a conceptual question regarding getMarkerFeatures() (and to make sure I'm understanding this function correctly):

I understand that getMarkerFeatures() as used below will return a list of markers that are unique to each group in the `groupBy` column data. If I understand correctly, the background group is set to all other cells not in the current group being examined.

```r
getMarkerFeatures(proj, groupBy = "myGroup")
```

However, if I run the below code instead - should I expect the same list of significant markers as compared to markers that are specific to Group_Level_1 from the above code? When I try it out on my data, I don't see the same list of significant markers. Any clarification on this would be greatly appreciated!

```r
getMarkerFeatures(proj, groupBy = "myGroup", useGroups = "Group_Level_1")
```

Thanks!

**Comment by rcorces** [comment]
Have you used `set.seed()` to set a constant seed? There are randomized operations in this function.

**Reply by wuv21** [reply]
Yes, I tried the two functions with the same seed (defined within the same scope as the two getMarkerFeature calls) and still see different results between the two.

**Reply by rcorces** [reply]
I'm not able to reproduce this behavior. Using the tutorial data I ran:

```
se1 <- getMarkerFeatures(projHeme5,
                        useMatrix = "GeneScoreMatrix", 
                        groupBy = "Clusters", threads = 16)

se2 <- getMarkerFeatures(projHeme5, useGroups = "C1",
                        useMatrix = "GeneScoreMatrix", 
                        groupBy = "Clusters", threads = 16)

se3 <- se1[,"C1"]

all.equal(rowData(se3)$name, rowData(se2)$name)

ggOneToOne(x = assays(se2)[["Log2FC"]][,1], y = assays(se3)[["Log2FC"]][,1])

all.equal(assays(se2)[["Log2FC"]][,1], assays(se3)[["Log2FC"]][,1])

```

both `all.equal()` calls return `TRUE` and the one to one plot shows that these are identical:
![image](https://user-images.githubusercontent.com/10343046/172463753-cecf0d5a-2554-48a6-a74f-30cefbab6362.png)

If you arent on the latest dev version of ArchR `release_1.0.2` then its possible that some previous code changes to how random operations are handled might be causing the behavior you're seeing.

**Reply by wuv21** [reply]
Thanks for the response!

To emulate your run, I also tried this out with release_1.0.2 at commit `cadcd31b6c228537d43701c63bd3ce9e7e1b2795` on the tutorial dataset and got identical results as you. Unfortunately for my own dataset, there was still a large discordance between the different addGroupCoverages. Any suggestions on why that might occur?

```r
se1 <- getMarkerFeatures(projOther,
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters2")

se2 <- getMarkerFeatures(projOther, useGroups = "C10",
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters2")

se3 <- se1[,"C10"]

all.equal(rowData(se3)$name, rowData(se2)$name)

ggOneToOne(x = assays(se2)[["Log2FC"]][,1], y = assays(se3)[["Log2FC"]][,1])

all.equal(assays(se2)[["Log2FC"]][,1], assays(se3)[["Log2FC"]][,1])
```

![image](https://user-images.githubusercontent.com/6870675/172623654-3a38bbd4-cfb3-44e6-9ea0-28bcbb34c190.png)

**Reply by rcorces** [reply]
> Unfortunately for my own dataset, there was still a large discordance between the different addGroupCoverages.

On your own dataset using `release_1.0.2`? Or an earlier release? If `release_1.0.2` then I'm not sure I have any intuition for why that might be.

**Reply by wuv21** [reply]
Yes with release_1.0.2 at commit cadcd31b6c228537d43701c63bd3ce9e7e1b2795. I'll debug through the different calls to see where the difference is happening and report back.

**Reply by wuv21** [reply]
The plot thickens! I found some more interesting things while playing with the tutorial data and my own data.

If I look at the second cluster in order (C2) for the tutorial data, it is no longer a straight correlation.

I tried out the first cluster (in sorted order) in my data and found that it was perfectly correlated. However, all other clusters were not correlated.

```r
se2 <- getMarkerFeatures(proj, useGroups = "C2",
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters")

se1 <- getMarkerFeatures(proj,
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters")

se3 <- se1[,"C2"]

all.equal(rowData(se3)$name, rowData(se2)$name) # this returns TRUE

ggOneToOne(x = assays(se2)[["Log2FC"]][,1], y = assays(se3)[["Log2FC"]][,1])
```
![image](https://user-images.githubusercontent.com/6870675/172739963-25f73e29-ee55-4c96-8b3a-5f0f63656520.png)

**Reply by wuv21** [reply]
My debugger is acting up right now, but I think something in this chunk is causing the discrepancy.

https://github.com/GreenleafLab/ArchR/blob/21099d6ed966f6088b7b49ebbb9cca07486e8875/R/MarkerFeatures.R#L641

**Reply by rcorces** [reply]
interesting. I am transferring this to an issue and will look into it.

**Reply by rcorces** [reply]
Moved to https://github.com/GreenleafLab/ArchR/issues/1467

**Reply by anastasiya-pendragon** [reply]
Hello there! Just wanted to give you an idea of why it works for the tutorial data and not in your dataset. 

EDIT: just now saw that you already tried it with a different cluster. 
So only going to leave this reference:
You can read my pull request here - [about markerFeatures potential bug](https://github.com/GreenleafLab/ArchR/pull/1409), there I explain where this difference comes from and how to potentially fix this. TLDR: it has to do with the random function that is used for the background cell selection. 
https://github.com/GreenleafLab/ArchR/blob/21099d6ed966f6088b7b49ebbb9cca07486e8875/R/MarkerFeatures.R#L675

The code in this line will have different random selections, depending on how many times it's called in a method. I.e. when you run your C10 cluster separately it's a different random seed than when you run it as the 10th cluster when you do not specify the usegroups. 

Hope it helps or at least clarifies some things! 
Best.
Anastasiya

**Reply by wuv21** [reply]
> Hello there! Just wanted to give you an idea of why it works for the tutorial data and not in your dataset.
> 
> EDIT: just now saw that you already tried it with a different cluster. So only going to leave this reference: You can read my pull request here - [about markerFeatures potential bug](https://github.com/GreenleafLab/ArchR/pull/1409), there I explain where this difference comes from and how to potentially fix this. TLDR: it has to do with the random function that is used for the background cell selection.
> 
> https://github.com/GreenleafLab/ArchR/blob/21099d6ed966f6088b7b49ebbb9cca07486e8875/R/MarkerFeatures.R#L675
> 
> The code in this line will have different random selections, depending on how many times it's called in a method. I.e. when you run your C10 cluster separately it's a different random seed than when you run it as the 10th cluster when you do not specify the usegroups.
> 
> Hope it helps or at least clarifies some things! Best. Anastasiya

Thanks @anastasiya-pendragon! Very helpful - will try it out and tag you in the issue that Ryan opened for this.
